socorro/lib/libjson.py

```diff
@@ -62,25 +63,230 @@ def lookup_definition(schema, ref):
+    new_schema = copy.deepcopy(lookup_definition(root_schema, ref))
+
+    # Copy anything from the original schema (other than "$ref") into the
+    # new schema
+    for key, val in schema.items():
+        if key == "$ref":
+            continue
+        new_schema[key] = copy.deepcopy(schema[key])
+
+    return new_schema
+
+
+def resolve_references(schema):
+    """Returns a schema with all ``$ref`` resolved
+
+    This works for both jsonschema schemas and socorro-data schemas.
+
+    .. Note::
+
+       When resolving a ``$ref``, all the properties are replaced with the exception of
+       "socorro" which is copied into the referenced schema.
+
+       For example::
+
+           "definitions": {
+               "color_definition": {
+                   "type": "object",
+                   "properties": {
+                       "name": {"type": "string"},
+                       "hex": {"type": "string"},
+                   },
+                   "foo": "bat",
+                   # This is overwritten by the referring schema
+                   "properties": ["public"],
+               },
+           },
+           "color": {
+               "$ref": "#/definitions/color_definition",
+               "foo": "bar",
+               "properties": ["protected"],
+           }
+
+       Results in:
+
+           "color": {
+               # From the referenced schema
+               "type": "object",
+               "properties": {
+                   "name": {"type": "string"},
+                   "hex": {"type": "string"},
+               },
+               "foo": "bat",
+               # From the referring schema
+               "properties": ["protected"],
+           }
 
     :arg dict schema: the schema
-    :arg dict schema_item: the sub-schema we're expanding a reference in
 
-    :returns: the new schema item with reference expanded
+    :returns: the new schema item with ``$ref`` resolved
 
     """
-    # FIXME(willkg): expanding references doesn't handle cycles, but we could handle
-    # that if we switched from recursion to iterative and maintaining context
-    if "$ref" in schema_item:
-        return lookup_definition(schema, schema_item["$ref"])
-    return schema_item
+    # FIXME(willkg): resolving references doesn't currently handle cycles
+
+    def _resolve(root_schema, schema):
+        if not schema:
+            return schema
+
+        if "$ref" in schema:
+            schema = resolve_reference(
+                root_schema=root_schema,
+                schema=schema,
+                ref=schema["$ref"],
+            )
+
+        type_ = schema["type"]
+        if "object" in type_:
+            if schema.get("properties"):
+                schema["properties"] = {
+                    key: _resolve(root_schema, val)
+                    for key, val in schema["properties"].items()
+                }
+            if schema.get("pattern_properties"):
```

Comment:
The docstring says this works for jsonschema and socorrodata schema, but it doesn't because jsonschema has `patternProperties` and socorrodata schema has `pattern_properties`.